ifrs17/Import/Importers.ipynb

```diff
@@ -541,6 +519,100 @@
+                "\n            ApplicationMessage.Log(Error.InactiveDataNodeState, item.First().DataNode);",
+                "\n}",
+                "\n"
+            ],
+            "metadata": {},
+            "execution_count": 0,
+            "outputs": []
+        },
+        {
+            "cell_type": "markdown",
+            "source": [
+                "## Validate for Data Node States Logic"
+            ],
+            "metadata": {},
+            "execution_count": 0,
+            "outputs": []
+        },
+        {
+            "cell_type": "code",
+            "source": [
+                "public async static Task ValidateDataNodeStatesAsync(this IWorkspace workspace, Dictionary<string, DataNodeData> persistentDataNodeByDataNode)",
+                "\n{",
+                "\n    foreach(var importedDataNodeState in await workspace.Query<DataNodeState>().ToArrayAsync())",
+                "\n    {",
+                "\n        if(persistentDataNodeByDataNode.TryGetValue(importedDataNodeState.DataNode, out var currentPersistentDataNode))",
+                "\n        {",
+                "\n            if(importedDataNodeState.State < currentPersistentDataNode.State)",
+                "\n                ApplicationMessage.Log(Error.ChangeDataNodeState, importedDataNodeState.DataNode, ",
+                "\n                                                                  currentPersistentDataNode.State.ToString(), ",
+                "\n                                                                  importedDataNodeState.State.ToString());",
+                "\n",
+                "\n            if(importedDataNodeState.State == currentPersistentDataNode.State)",
+                "\n                await workspace.DeleteAsync<DataNodeState>(importedDataNodeState);",
+                "\n        }",
+                "\n    }",
+                "\n}"
+            ],
+            "metadata": {},
+            "execution_count": 0,
+            "outputs": []
+        },
+        {
+            "cell_type": "markdown",
+            "source": [
+                "## Validate for Mandatory Aoc Steps"
+            ],
+            "metadata": {},
+            "execution_count": 0,
+            "outputs": []
+        },
+        {
+            "cell_type": "code",
+            "source": [
+                "public async static Task ValidateForMandatoryAocSteps(this IWorkspace workspace, IDataSet dataSet, HashSet<AocStep> mandatoryAocSteps)",
+                "\n{",
+                "\n    var includedProperties = typeof(RawVariable).GetIdentityProperties().Select(x=>x.Name).Except(new[]{nameof(AocType), nameof(Novelty)});",
+                "\n    var missingAocStepsByIdentityProperties = (await workspace.Query<RawVariable>().ToListAsync())",
+                "\n                     .GroupBy(x => string.Join(\", \", x.ToString().Split('{','}')[1].Split(',').Where(y=> includedProperties.Contains(y.Split('=')[0].Trim()))).Trim(),",
```

Comment:
what is this line doing? can we simplify it?

Reply:
-the following. if i set:
var x = new RawVariable{DataNode = "A", EstimateType = "B"};
-then
x.ToString()
-prints: 
RawVariable { DataNode = A, AocType = , Novelty = , Id = 00000000-0000-0000-0000-000000000000, Partition = 00000000-0000-0000-0000-000000000000, AmountType = , AccidentYear = , Values = , EstimateType = B }
-then the long line takes:
DataNode = A, AmountType = , AccidentYear = , EstimateType = B

Reply:
Can't just use x.Select() instead of the splits?

Reply:
![image](https://user-images.githubusercontent.com/120557622/212285726-e6879922-c8b8-43c1-82af-e736d9dd2b12.png)

Reply:
Select works if one knows all the properties they want to select. The goal of the method was to work also if the data model changes. Then I did not know how to make it work with the Select (or equivalent in the GroupBy clause).

Reply:
there is a ToIdentityString() method. What about this one?

Reply:
Good proposition.
I included it in the new version.